pallets/parachain-staking/src/types.rs

```diff
@@ -0,0 +1,1673 @@
+                owner: collator,
+                amount,
+            }]),
+            total: amount,
+            less_total: Balance::zero(),
+            status: DelegatorStatus::Active,
+        }
+    }
+
+    pub fn default_with_total(id: AccountId, amount: Balance) -> Self {
+        Delegator {
+            id,
+            total: amount,
+            delegations: OrderedSet::from(vec![]),
+            less_total: Balance::zero(),
+            status: DelegatorStatus::Active,
+        }
+    }
+
+    pub fn total(&self) -> Balance {
+        self.total
+    }
+
+    pub fn total_sub_if<T, F>(&mut self, amount: Balance, check: F) -> DispatchResult
+    where
+        T: Config,
+        T::AccountId: From<AccountId>,
+        BalanceOf<T>: From<Balance>,
+        F: Fn(Balance) -> DispatchResult,
+    {
+        let total = self.total.saturating_sub(amount);
+        check(total)?;
+        self.total = total;
+        self.adjust_bond_lock::<T>(BondAdjust::Decrease)?;
+        Ok(())
+    }
+
+    pub fn total_add<T, F>(&mut self, amount: Balance) -> DispatchResult
+    where
+        T: Config,
+        T::AccountId: From<AccountId>,
+        BalanceOf<T>: From<Balance>,
+    {
+        self.total = self.total.saturating_add(amount);
+        self.adjust_bond_lock::<T>(BondAdjust::Increase(amount))?;
+        Ok(())
+    }
+
+    pub fn total_sub<T>(&mut self, amount: Balance) -> DispatchResult
+    where
+        T: Config,
+        T::AccountId: From<AccountId>,
+        BalanceOf<T>: From<Balance>,
+    {
+        self.total = self.total.saturating_sub(amount);
+        self.adjust_bond_lock::<T>(BondAdjust::Decrease)?;
+        Ok(())
+    }
+
+    pub fn is_active(&self) -> bool {
+        matches!(self.status, DelegatorStatus::Active)
+    }
+
+    pub fn add_delegation(&mut self, bond: Bond<AccountId, Balance>) -> bool {
+        let amt = bond.amount;
+        if self.delegations.insert(bond) {
+            self.total = self.total.saturating_add(amt);
+            true
+        } else {
+            false
+        }
+    }
+    // Return Some(remaining balance), must be more than MinDelegatorStk
+    // Return None if delegation not found
+    pub fn rm_delegation<T: Config>(&mut self, collator: &AccountId) -> Option<Balance>
+    where
+        BalanceOf<T>: From<Balance>,
+        T::AccountId: From<AccountId>,
+    {
+        let mut amt: Option<Balance> = None;
+        let delegations = self
+            .delegations
+            .0
+            .iter()
+            .filter_map(|x| {
+                if &x.owner == collator {
+                    amt = Some(x.amount);
+                    None
+                } else {
+                    Some(x.clone())
+                }
+            })
+            .collect();
+        if let Some(balance) = amt {
+            self.delegations = OrderedSet::from(delegations);
+            self.total_sub::<T>(balance)
+                .expect("Decreasing lock cannot fail, qed");
```

Comment:
expect

Reply:
total_sub uses saturating_sub so it really can not fail even if balance her is provided too large